server/trillian_log_server.go

```diff
@@ -18,17 +19,47 @@ import (
 // Pass this as a fixed value to proof calculations. It's used as the max depth of the tree
 const proofMaxBitLen = 64
 
-// LogStorageProviderFunc decouples the server from storage implementations
+// LogStorageProviderFunc generates a LogStorage instance for a given log ID; it is a
+// separate type in order to decouple the server from storage implementations.
 type LogStorageProviderFunc func(int64) (storage.LogStorage, error)
 
+// cachedLogStorageProvider wraps a LogStorageProviderFunc and caches its results.
+type cachedLogStorageProvider struct {
+	provider    LogStorageProviderFunc
+	mu          sync.Mutex
+	logStorages map[int64]storage.LogStorage // guarded by mu
+}
+
+// newCachedLogStorageProvider creates a new cached storage provider for the given
+// raw provider.
+func newCachedLogStorageProvider(p LogStorageProviderFunc) cachedLogStorageProvider {
+	return cachedLogStorageProvider{provider: p, logStorages: make(map[int64]storage.LogStorage)}
+}
+
+func (c *cachedLogStorageProvider) storageForLog(logID int64) (storage.LogStorage, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	s, ok := c.logStorages[logID]
+	if ok {
+		return s, nil
+	}
+
+	s, err := c.provider(logID)
+	if err == nil {
```

Comment:
Maybe flip this and return err early, and then store + return below?
(indent error flow)

```go
s, err := ...
if err != nil {
   return nil, err
}
c.logStorages[logID] = s
return s, nil
```